src/Data/Singletons/Promote.hs

```diff
@@ -750,53 +815,96 @@ promoteLetDecRHS rhs_sort type_env fix_env mb_let_uniq name let_dec_rhs = do
     -- * For UFunctions, `prom_a` is [DTySynEqn] and `a` is [DClause].
     promote_let_dec_rhs
       :: [Name]                            -- Local variables bound in this scope
-      -> [Maybe DKind]                     -- Promoted function argument types (if known)
-      -> Maybe DKind                       -- Promoted function result type (if known)
+      -> Maybe LetDecRHSKindInfo           -- Information about the promoted kind (if present)
       -> Int                               -- The number of promoted function arguments
       -> PrM (prom_a, a)                   -- Promote the RHS
       -> (prom_a -> [DTySynEqn])           -- Turn the promoted RHS into type family equations
       -> (DType -> Int -> a -> ALetDecRHS) -- Build an ALetDecRHS
-      -> PrM (DDec, [DDec], ALetDecRHS)
-    promote_let_dec_rhs all_locals m_argKs m_resK ty_num_args
+      -> PrM (DDec, Maybe DDec, [DDec], ALetDecRHS)
+    promote_let_dec_rhs all_locals m_ldrki ty_num_args
                         promote_thing mk_prom_eqns mk_alet_dec_rhs = do
       opts <- getOptions
-      tyvarNames <- mapM (const $ qNewName "a") m_argKs
+      tyvarNames <- replicateM ty_num_args (qNewName "a")
       let proName         = promotedValueName opts name mb_let_uniq
           local_tvbs      = map DPlainTV all_locals
-          m_fixity        = OMap.lookup name fix_env
-          args            = zipWith inferMaybeKindTV tyvarNames m_argKs
+          args            = case m_ldrki of
```

Comment:
Too many matches on `Maybe`s here. Could this just be done once, with `(a, b, c, d) | Just (LDRKI ...) <-... = ... | otherwise = ...`?

Reply:
Funny enough, I chose this design because I thought `(a, b, c, d) | Just (LDRKI ...) <-... = ... | otherwise = ...` looked uglier than the current design (which, admittedly, is not much prettier in my eyes). I've changed the code back to your suggested design—is this to your liking?

Reply:
Maybe I'm wrong, but I think we can do even better. There are several other definitions that immediately scrutinize `Maybe`s. Like `m_sak_dec`, `defun_ki`, and `tf_head`.

Reply:
I've factored out the definitions of `mk_sak_dec`, `defun_ki`, and `tf_head` to use a single `case`.

Reply:
What we really have here is three cases:
1. We have no information.
2. We have a `LetDecRHSKindInfo`, but no SAK.
3. We have a `LetDecRHSKindInfo` with a SAK.

The three cases lead to slightly different algorithms. But this fact is not currently all that apparent from the code. My hope is that there would be one `case`, with three branches, that distinguishes the three cases. That said, perhaps that would cause other trouble in the code. I don't want this to hold up the patch -- just trying to make the code paths more apparent.

Reply:
Ah, now I see what you're getting at. Indeed, that is a nice way to structure this code, and thinking about it that way made me realize that some of the things returned from the first tuple-returning `case` only exist to be fed into the second tuple-returning `case`. Combining both these `case`s into one allowed me to reduce the number of tupled return values from seven to four, which is much nicer.

I took some liberties with your suggestion and numbered the cases as 1/2(a)/2(b) instead of 1/2/3, since the last two cases share a few commonalities (e.g., they compute the exact same `lde_kvs_to_bind`). How does this look now?

Reply:
I haven't head any objections by now, so I'll interpret that to mean that there are no further suggestions to this code. Let's land this.

Reply:
I hadn't had time to look over the past few days, but yes I'm quite happy with this now. Thanks!